Optimisation on grid data fetching

SianaDicheva opened the pull request:
[Signal Benchmarking.xlsx](https://github.com/ReliefApplications/ems-backend/files/14178950/Signal.Benchmarking.xlsx)
# Description

Introducing caching on user permissions. As user permissions do not get updated very often but take time getting queried when fetching records for grid widgets, we are caching them for 10 minute periods to optimise the performance of system.

The main change in code is in **all.ts** where the caching is implemented. 

## Useful links

- Please insert link to ticket - https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/68879

## Type of change

Please delete options that are not relevant.

- [x] Improvement (refactor or addition to existing functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B


# Checklist:

( * == Mandatory ) 

- [ ] * I have set myself as assignee of the pull request
- [x] * My code follows the style guidelines of this project
- [x] * Linting does not generate new warnings
- [ ] * I have performed a self-review of my own code
- [ ] * I have put the ticket for review, adding the oort-backend team to the list of reviewers
- [x] * I have commented my code, particularly in hard-to-understand areas
- [ ] * I have put JSDoc comment in all required places
- [x] * My changes generate no new warnings
- [ ] * I have included screenshots describing my changes if relevant
- [ ] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

![image](https://github.com/ReliefApplications/ems-backend/assets/78383370/838fa2f7-9d72-4cce-a99c-578e412bddcd)
![image](https://github.com/ReliefApplications/ems-backend/assets/78383370/fac8ca5b-0a85-48b2-8189-6480002e37bc)
![image](https://github.com/ReliefApplications/ems-backend/assets/78383370/201cc0e8-9db7-4cdf-9fc8-b1867e5b39be)

AntoineRelief commented:
@SianaDicheva 
Thanks for the pull request!
Just got a few questions / queries
- Is it the final code?
- Can you please add more description to the PR, filling the different fields of the template
- Is there any metrics / tests that would indicate how it is affecting performance?

SianaDicheva commented:
> @SianaDicheva Thanks for the pull request! Just got a few questions / queries
> 
> * Is it the final code?
> * Can you please add more description to the PR, filling the different fields of the template
> * Is there any metrics / tests that would indicate how it is affecting performance?

Hey Antoine, Yes, this is the final code. I believe I added description and filled the necessary fields. I performed some tests and attached an excel spreadsheet with performance metrics and attached it to PR. I also added couple of screenshots as it is more visual.
Hope this clears things up!
Please let me know if you have any more comments or concerns :)

AntoineRelief commented:
hi @SianaDicheva 
thanks for the update!
a couple of additional requests from my side:
- the benchmark does not indicate previous loading time. It would be interesting to compare with previous performance
- in which environment were the tests made?
- I can see jsdoc comments were added to the code, but they are all empty. I think this is because a few additional lines were added, and you used the lint --fix to add them. Similar comment was made by Matheus on another PR. Please check the comments before validating the PRs, in order to avoid empty ones / useless ones

AntoineRelief commented on src/utils/schema/resolvers/Query/all.ts:
```diff
@@ -703,7 +735,7 @@ export default (entityName: string, fieldsByName: any, idsByName: any) =>
       // === CONSTRUCT OUTPUT + RETURN ===
       const edges = items.map((r) => {
-        const record = getAccessibleFields(r, ability);
+        const record = getAccessibleFields(r, abilityCache.get(userId));
```
@SianaDicheva 
it's really an edge case but there is a possibility there that the cache is cleared, between the time you add the user ability to the cache, and the time you get the accessible fields
as the cache is automatically cleared every 10minutes, regardless of when you added a new user or form to it, this situation may happen

AntoineRelief replied:
@SianaDicheva 
I'll make some changes to the PR, please don't push any new code

AntoineRelief commented:
@SianaDicheva 
closing the PR, and did the update on this one:
https://github.com/ReliefApplications/ems-backend/pull/972

I struggled to make changes directly on the branch of the PR, so I found this alternative

A few changes I made:
- remove the searchUsed boolean in the getSearchFilter file, and prefer to use a test based on the length of the search stage
- changed the searchStage so it's not declared at file level, but inside the method
- there was a mistake in the all.ts file. Indeed, the ability should have been calculated on all forms, instead of a single form ( due to the fact that the query can create links between forms ). I noticed this difference while comparing to the download method. It was out of scope from your PR, but I took the opportunity to fix it
